File: python/dfa.py
class DFA:
    def __init__(self, initial_state, accept_states, transition):
        self.initial_state = initial_state
        self.accept_states = set(accept_states)
        self.transition = transition
        self.current_state = initial_state

    def reset(self):
        self.current_state = self.initial_state

    def process(self, symbol):
        if symbol in self.transition[self.current_state]:
            self.current_state = self.transition[self.current_state][symbol]
        else:
            # If no valid transition, move to an error state
            self.current_state = None

    def accept(self):
        return self.current_state in self.accept_states

    def process_string(self, string):
        self.reset()
        for symbol in string:
            self.process(symbol)
            if self.current_state is None:
                return False
        return self.accept()


# Define transitions for the DFA
transition = {
    'q0': {'0': 'q1', '1': 'q0'},
    'q1': {'0': 'q0', '1': 'q1'},
}

# Create a DFA with initial state 'q0' and accept state 'q0'
dfa = DFA('q0', ['q0'], transition)

# Test strings
test_strings = [
    "0",
    "1",
    "00",
    "01",
    "10",
    "11",
    "000",
    "001",
    "010",
    "011",
    "100",
    "101",
    "110",
    "111",
]

for s in test_strings:
    print(f"String: {s}, Accepted: {dfa.process_string(s)}")

File: go/DFA/dfa.go
///*  DFA, which accepts binary strings containing an even number of 0s: */

package main

import "fmt"

// DFA struct representing a deterministic finite automaton
type DFA struct {
	initialState int
	acceptStates map[int]bool
	transition   map[int]map[rune]int
	currentState int
}

// NewDFA initializes and returns a new DFA
func NewDFA(initialState int, acceptStates []int, transition map[int]map[rune]int) *DFA {
	acceptStatesMap := make(map[int]bool)
	for _, state := range acceptStates {
		acceptStatesMap[state] = true
	}

	return &DFA{
		initialState: initialState,
		acceptStates: acceptStatesMap,
		transition:   transition,
		currentState: initialState,
	}
}

// Reset resets the DFA to its initial state
func (d *DFA) Reset() {
	d.currentState = d.initialState
}

// Process processes a single character and transitions to the next state
func (d *DFA) Process(c rune) {
	if nextState, ok := d.transition[d.currentState][c]; ok {
		d.currentState = nextState
	} else {
		// No valid transition, stay in current state (or handle as needed)
		d.currentState = -1 // Invalid state
	}
}

// Accept checks if the DFA is in an accept state
func (d *DFA) Accept() bool {
	return d.acceptStates[d.currentState]
}

// ProcessString processes an entire string through the DFA
func (d *DFA) ProcessString(s string) bool {
	d.Reset()
	for _, c := range s {
		d.Process(c)
		if d.currentState == -1 {
			return false
		}
	}
	return d.Accept()
}

func main() {
	// Define transitions for the DFA
	transition := map[int]map[rune]int{
		0: {'0': 1, '1': 0},
		1: {'0': 0, '1': 1},
	}

	// Create a DFA with initial state 0, accept state 0
	dfa := NewDFA(0, []int{0}, transition)

	// Test strings
	testStrings := []string{
		"0",
		"1",
		"00",
		"01",
		"10",
		"11",
		"000",
		"001",
	}

	for _, s := range testStrings {
		fmt.Printf("String: %s, Accepted: %t\n", s, dfa.ProcessString(s))
	}
}
